### DISCUSSION TITLE
Transitions with Short Items (Station ID)

### DISCUSSION THREAD

**Discussion by Pelican104** [savonet/liquidsoap#2376] [User Support]
I've been using Liquidsoap for several years now to run speech based radio stations and it works brilliantly.

I'd also like to use LS to playout some music based stations but I am being held back by the inability to figure out how to get simple three way transitions to work. These transitions are the most common in all my music services but I just cant make them work, and really they are nothing complicated.

I have LS set to use request.dynamic to get the next track from my external scheduling system. that works well and the requests are annotated with liq_start_next metadata to indicate when the next track should start.

The problem is that most transitions take the format SONG A > STATION ID > SONG B

The station ID's tend to be short (3-6 seconds). 

What happens is that SONG A reaches it's liq_Start_next trigger and the STATION ID starts as expected. But then because the STATION ID is short, the tail end of SONG A is still playing when the STATION ID reaches it's liq_start_next trigger. This trigger then seems to get ignored and SONG B doesn't start until SONG A has completely finished.

What am I missing? I see lots of people running music stations, how do you get this to work?

I'm currently using 1.41 on Ubuntu.

![SONG A](https://user-images.githubusercontent.com/42071538/165836596-44cff9ef-b38c-4a5e-9355-0cb2f39f8657.png)

**Comment by bjarn** [comment]
I was also wondering how to do this, I remember @toots and I had some brainstorm sessions about this on Slack almost a year ago. Not sure what the options are now LS2 has been released?

For now; I have a cross duration of max. 1.5 seconds which matches the shortest station ID and set a cue out timestamp on the previous track + a small fadeout if needed. It then sounds "okay". But, overlapping definitely sounds better than this workaround.

**Comment by toots** [comment]
Hi! could you describe how this is implemented in your script a little bit more? Thanks!

**Reply by Pelican104** [reply]
Hello there, here is the script.....

I've got a request function which is getting the media URI's from an external scheduler. The request function also receives the START NEXT marker for the track  from the scheduling system (called OUTRO) which it annotates into the request.

I'm then trying to use CROSSFADE to achieve the transitions. It works fine from SONG A in to the ID but then SONG B doesn't start until SONG A has completely finished. 


#! /usr/bin/liquidsoap

set("init.allow_root",true)
set("server.telnet",true)
set("harbor.ssl.certificate","REDACTED")
set("harbor.ssl.private_key","REDACTED")
set("scheduler.fast_queues",2)
set("scheduler.generic_queues",3)

live = input.harbor.ssl("live",port=8080,user="REDACTED",password="REDACTED",buffer=1.0,max=2.0)
logid = ref("")

s=sine()

# Functions
def get_request() =
  uri = "REDACTED.php?logid=" ^ !logid
  print(uri)
  let (st,hd,data) = http.get(uri)
  print(st)
  print(hd)
  
  myjson = of_json(default= [("error","fail")], string_of(data))
  
  #if it's an advert then find the audio in the advert path
  if myjson['spot'] == "1" then 
   filenm = 'annotate:liq_cross_duration="' ^ myjson["outro"] ^ '",liq_fade_in="0.0":REDACTED_URL' ^ myjson["media"]
   print(filenm)
   logid := string_of(myjson['logid'])
   request.create(filenm, persistent=true) 
  else
   #if it's not an advert find the audio in the main audio store
   filenm = 'annotate:liq_cross_duration="' ^ myjson["outro"] ^ '",liq_start_next="' ^ myjson["outro"] ^ '",liq_fade_in="0.0":REDACTED_URL' ^ myjson["media"]
   print(filenm)
   logid := string_of(myjson['logid'])
   request.create(filenm, persistent=true)
  end  
end


# Requests
requests = request.dynamic(
  id="requests",
  conservative=false,
  get_request
)

radio = fallback(track_sensitive=false,[live,requests,s])
radio = crossfade(radio)
radio = sky(radio)

# Stream it out
output.icecast(%vorbis,
  host = "localhost", port = 8000,
  password = "REDACTED", mount = "radio.ogg",
  radio)

**Reply by bjarn** [reply]
For every track, I set the metadata like this:
```
annotate:title="West Coast",artist="OneRepublic",duration="192.94",song_id="241c83ccf9b519c8edca1e46508ed75f",media_id="6964",liq_amplify="-3.50dB",liq_cross_duration="1.80",liq_cue_in="0.00",liq_cue_out="192.90",playlist_id="56":media:Muziek/onerepublic_-_west_coast.mp3
```

Especially the `liq_cross_duration` and `liq_cue_out` are important here.

Then, I have this transition script:
```
# Crossfade --------------------->

def crossfade(~duration=5.,~override_duration="liq_cross_duration",
              ~fade_in=3.,~fade_out=3.,~smart=false,
              ~default=(fun (a,b) -> (sequence([a, b]):source)),
              ~high=-15., ~medium=-32., ~margin=4.,
              ~minimum=(-1.),~width=2.,~conservative=true, s)
  let fade.out = fade.out(type="sin",duration=fade_out)
  let fade.in  = fade.in(type="sin",duration=fade_in)

  add = fun (a,b) -> add(normalize=false,[b, a])
  log = log(label="crossfade")

  def simple_transition(a, b)
    list.iter(fun(x)-> log(level=4,"Before: #{x}"), a.metadata)
    list.iter(fun(x)-> log(level=4,"After : #{x}"), b.metadata)

    log("Simple transition: crossed, fade-in, fade-out.")

    // We set the crossfade duration here, which for the track above is 1.8 seconds.
    cross_duration = float_of_string(default=duration, a.metadata[override_duration])
    after_duration = float_of_string(default=(-1.), b.metadata["duration"])

    log("cross duration: #{cross_duration}, after duration: #{after_duration}")

    sa = fade.out(a.source)
    sb = fade.in(b.source)

    sb =
      if after_duration > 0. and after_duration < cross_duration then
        log("Padding with #{cross_duration-after_duration} of silence")
        sequence([blank(duration=cross_duration-after_duration), sb])
      else
        sb
      end

    add(sa, sb)
  end

  cross(width=width, duration=duration, conservative=conservative,
        minimum=minimum, simple_transition,s)
end

# Finalize stream ------------->

// Default fade_out and fade_in are .01 seconds.
radio = crossfade(fade_out=.01, fade_in=.01, default=(fun (a,b) -> add(normalize=false,[b, a])), conservative=true,  radio)
```

However, this script is not working as of 2.0.2+ due to an issue mentioned in https://github.com/savonet/liquidsoap/issues/2370

The padding with blank also has not the desired effects, as the next track should just start earlier and the station ID should not have blank added.

That's what you also want right? @Pelican104

**Reply by Pelican104** [reply]
That's interesting - I will have a play with that, not sure I entirely understand how putting silence on the start of the B item helps, but I will try it and see.

Thanks

**Reply by bjarn** [reply]
Good luck. The silence part doesn't really help indeed, it never gets triggered for me as far as I have seen, but that might be because of the 1.5 seconds I set as minimum for my crosses.

But it's just a workaround, your proposal is still something I am looking for too.
